File: roblox/src/shared/Stores.lua
-- Store definitions (vendor trade lists). Source of truth is
-- backend/content/stores.json, overlaid at boot from GET /content the same
-- way as Items (see ContentService/ContentSync). The static table below is
-- the fallback for Studio-without-HTTP and backend outages.
--
-- Prices are gold per unit; a trade may be buy-only (no sellPrice),
-- sell-only (no buyPrice), or both. Vendor NPC placement lives in
-- server/VendorService — this module is economy data only.

-- `warn` is a Roblox global; the fallback keeps this module runnable under
-- the headless Luau CLI (content overlay tests).
local warn = warn or print

local Stores = {}

Stores.defs = {
	general_goods = {
		id = "general_goods",
		name = "General Goods",
		trades = {
			{ itemId = "sword_iron", buyPrice = 120, sellPrice = 40 },
			{ itemId = "helmet_leather", buyPrice = 40, sellPrice = 12 },
			{ itemId = "chest_leather", buyPrice = 60, sellPrice = 18 },
			{ itemId = "gloves_leather", buyPrice = 30, sellPrice = 9 },
			{ itemId = "legs_leather", buyPrice = 45, sellPrice = 13 },
			{ itemId = "boots_leather", buyPrice = 30, sellPrice = 9 },
			{ itemId = "ring_vitality", buyPrice = 150, sellPrice = 50 },
			{ itemId = "ring_focus", buyPrice = 150, sellPrice = 50 },
			{ itemId = "wood", sellPrice = 2 },
			{ itemId = "stone", sellPrice = 2 },
			{ itemId = "slime_goo", sellPrice = 3 },
			{ itemId = "goblin_ear", sellPrice = 5 },
		},
	},
}

function Stores.get(storeId)
	return Stores.defs[storeId]
end

-- The trade entry for an item in a store, or nil if the store doesn't
-- carry it.
function Stores.trade(storeId, itemId)
	local store = Stores.defs[storeId]
	if not store then
		return nil
	end
	for _, trade in ipairs(store.trades) do
		if trade.itemId == itemId then
			return trade
		end
	end
	return nil
end

-- Overlay backend-served store defs, same contract as Items.apply: fetched
-- stores replace local ones wholesale, local-only stores survive but get
-- flagged as mirror drift. Returns true if applied.
function Stores.apply(storeDefs)
	if type(storeDefs) ~= "table" then
		return false
	end

	local fetched = {}
	for id, def in pairs(storeDefs) do
		Stores.defs[id] = def
		fetched[id] = true
	end

	local localOnly = {}
	for id in pairs(Stores.defs) do
		if not fetched[id] then
			table.insert(localOnly, id)
		end
	end
	if #localOnly > 0 then
		warn(
			"[Stores] defs only in the Luau mirror, missing from backend content: "
				.. table.concat(localOnly, ", ")
		)
	end

	return true
end

return Stores

File: backend/src/stores.js
// Store definitions (vendor trade lists), loaded from the git-tracked
// content file (content/stores.json). Prices are gold per unit; an entry
// may be buy-only (no sellPrice), sell-only (no buyPrice), or both.
// Vendor NPC placement lives Roblox-side (VendorService VENDOR_DEFS) —
// the backend owns the economy data, the game owns the world layout.
//
// Like items.js, validation fails the boot loudly on malformed content.

import fs from "node:fs";
import { ITEMS } from "./items.js";

const raw = JSON.parse(
  fs.readFileSync(new URL("../content/stores.json", import.meta.url), "utf8")
);

function fail(message) {
  throw new Error(`content/stores.json: ${message}`);
}

function validatePrice(where, value) {
  if (value !== undefined && (!Number.isInteger(value) || value < 1)) {
    fail(`${where} must be a positive integer`);
  }
}

for (const [key, store] of Object.entries(raw.stores)) {
  if (store.id !== key) fail(`store "${key}" has mismatched id "${store.id}"`);
  if (typeof store.name !== "string" || !store.name) fail(`store "${key}" needs a name`);
  if (!Array.isArray(store.trades) || store.trades.length === 0) {
    fail(`store "${key}" needs a non-empty trades list`);
  }
  const seen = new Set();
  for (const trade of store.trades) {
    const where = `store "${key}" trade "${trade.itemId}"`;
    if (!ITEMS[trade.itemId]) fail(`${where} references an unknown item`);
    if (seen.has(trade.itemId)) fail(`${where} is listed twice`);
    seen.add(trade.itemId);
    validatePrice(`${where} buyPrice`, trade.buyPrice);
    validatePrice(`${where} sellPrice`, trade.sellPrice);
    if (trade.buyPrice === undefined && trade.sellPrice === undefined) {
      fail(`${where} needs a buyPrice and/or sellPrice`);
    }
  }
}

export const STORES = raw.stores;
